test/testdrive/storage-clusters.td

```diff
@@ -0,0 +1,170 @@
+contains:cannot create more than one replica of a cluster containing sources or sinks
+
+# Creating indexes and materialized views in a cluster is permitted once the
+# source has been dropped.
+> DROP SOURCE loadgen;
+> CREATE INDEX idx IN CLUSTER storage ON t (a)
+> CREATE MATERIALIZED VIEW mv IN CLUSTER storage AS SELECT 1
+
+# Recreating the source is banned.
+! CREATE SOURCE loadgen IN CLUSTER storage FROM LOAD GENERATOR COUNTER
+contains:cannot create source in cluster containing indexes or materialized views
+
+# Creating a second replica is now allowed.
+> CREATE CLUSTER REPLICA storage.r2 SIZE '1'
+
+# The source cannot be readded even after the index and materialized views
+# are dropped because of the second replica.
+> DROP INDEX idx
+> DROP MATERIALIZED VIEW mv
+! CREATE SOURCE loadgen IN CLUSTER storage FROM LOAD GENERATOR COUNTER
+contains:cannot create source in cluster with more than one replica
+
+# Creating the source is allowed after dropping the second replica
+> DROP CLUSTER REPLICA storage.r2
+> CREATE SOURCE loadgen IN CLUSTER storage FROM LOAD GENERATOR COUNTER
+
+# Test that `DROP CLUSTER` only succeeds with `CASCADE`.
+! DROP CLUSTER storage
+contains:cannot drop cluster with active objects
+> DROP CLUSTER storage CASCADE
+
+# Test that a cluster can contain multiple sources and sinks, and verify that
+# the sources and sinks produce the correct output.
+
+> CREATE CLUSTER storage REPLICAS (r1 (SIZE '1'))
+
+$ kafka-create-topic topic=data1 partitions=1
+
+$ kafka-create-topic topic=data2 partitions=1
+
+$ kafka-ingest format=bytes topic=data1
+a
+b
+
+$ kafka-ingest format=bytes topic=data2
+aa
+bb
+
+> CREATE CONNECTION kafka
+  TO KAFKA (BROKER '${testdrive.kafka-addr}')
+
+> CREATE SOURCE data1
+  IN CLUSTER storage
+  FROM KAFKA CONNECTION kafka (TOPIC 'testdrive-data1-${testdrive.seed}')
+  FORMAT TEXT
+
+> CREATE SOURCE data2
+  IN CLUSTER storage
+  FROM KAFKA CONNECTION kafka (TOPIC 'testdrive-data2-${testdrive.seed}')
+  FORMAT TEXT
+
+> SELECT * FROM data1
+a
+b
+
+> SELECT * FROM data2
+aa
+bb
+
+> CREATE MATERIALIZED VIEW view1 AS SELECT text FROM data1
+
+> CREATE MATERIALIZED VIEW view2 AS SELECT text || text AS text FROM data2
+
+> CREATE SINK sink1
+  IN CLUSTER storage
+  FROM view1
+  INTO KAFKA CONNECTION kafka (TOPIC 'sink1-${testdrive.seed}')
+  FORMAT JSON ENVELOPE DEBEZIUM
+
+> CREATE SINK sink2
+  IN CLUSTER storage
+  FROM view2
+  INTO KAFKA CONNECTION kafka (TOPIC 'sink2-${testdrive.seed}')
+  FORMAT JSON ENVELOPE DEBEZIUM
+
+$ kafka-verify-data format=json sink=materialize.public.sink1 key=false sort-messages=true
+{"before": null, "after": {"text": "a"}}
+{"before": null, "after": {"text": "b"}}
+
+$ kafka-verify-data format=json sink=materialize.public.sink2 key=false sort-messages=true
+{"before": null, "after": {"text": "aaaa"}}
+{"before": null, "after": {"text": "bbbb"}}
+
+# Test that the replica can be sized up and the sources and sinks correctly
```

Comment:
cc @philip-stoev you might want to add more structured tests that exercise this behaviour?